File: src/image_shrink_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import roslib; roslib.load_manifest('image_shrink')
import rospy as r
import cv2
import numpy as np
from tempfile import TemporaryFile
from sensor_msgs.msg import Image
from image_shrink.msg import StringArray
from cv_bridge import CvBridge, CvBridgeError

class ImageShrinkClient:
    def __init__(self):
        r.init_node('image_shrink_client')
        self.pub = r.Publisher('out_image', Image)
        self.sub = r.Subscriber('image_shrinked', StringArray, self.callback)
        self.bridge = CvBridge()

    def run(self):
        r.Rate(r.get_param('~rate'))
        r.spin()

    def decompressMessage(self, src):
        height = int(src[0])
        width = int(src[1])
        tmp = TemporaryFile()
        tmp.writelines(src[2:])
        tmp.seek(0)
        npz = np.load(tmp)
        body = npz['img']
        ret = np.ndarray((height, width), dtype=np.uint8)
        for i in range(height*width):
            if body[i]:
                ret[i/width][i%width] = 255


    def callback(self, sarray):
        height = sarray[0]
        width = sarray[1]
        tmp = TemporaryFile()
        tmp.writelines(sarray.data)
        tmp.seek(0)

        data = np.load(tmp)
        img = data['img']
        img.shape = (img.shape[0], img.shape[1], 1)
        imgmsg = self.bridge.cv2_to_imgmsg(img)
        imgmsg.encoding = 'mono8'
        self.pub.publish(imgmsg)

def init():
    shrink_client = ImageShrinkClient()
    shrink_client.run()

if __name__ == '__main__':
    init()

File: src/image_shrink_client.cpp
#include <boost/shared_ptr.hpp>
#include <vector>
#include <iostream>
#include <opencv2/imgproc/imgproc.hpp>
#include <cv_bridge/cv_bridge.h>

#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/image_encodings.h>
#include <image_shrink/BinaryImage.h>

namespace enc = sensor_msgs::image_encodings;

class ImageShrinkClient {
    ros::NodeHandle _n;
    ros::NodeHandle _ln;
    ros::Publisher _pub;
    ros::Subscriber _sub;
    sensor_msgs::ImagePtr _img_ptr;
    bool _isInitialized;

public:
    ImageShrinkClient() : _ln(ros::NodeHandle("~")), _isInitialized(false) {
        _pub = _n.advertise<sensor_msgs::Image>("image_edged", 10);
        _sub = _n.subscribe<image_shrink::BinaryImage>("image_binary", 1, &ImageShrinkClient::binaryCallback, this);
    }
    virtual ~ImageShrinkClient(){};

    void initialize(){
        _img_ptr = boost::make_shared<sensor_msgs::Image>();
    }

    void binaryCallback(const image_shrink::BinaryImageConstPtr& msg){
        if (!_isInitialized){
            initialize();
            _isInitialized = true;
        }
        _img_ptr->height = msg->height;
        _img_ptr->width = msg->width;
        _img_ptr->encoding = "mono8";
        _img_ptr->step = msg->width;
        _img_ptr->data.resize(_img_ptr->height * _img_ptr->width);
        for (int y = 0; y < (int)_img_ptr->height; ++y){
            for (int x = 0; x < (int)_img_ptr->width; ++x){
                int i = y * _img_ptr->width + x;
                if (((msg->data[i/8] >> (i%8)) & 1) == 1){
                    _img_ptr->data[i] = 255;
                } else {
                    _img_ptr->data[i] = 0;
                }
            }
        }
        _pub.publish(_img_ptr);
    } // end of callback
};

int main(int argc, char *argv[])
{
    ros::init(argc, argv, "image_shrink_client");
    ImageShrinkClient client;
    ros::Rate r(10);
    while (ros::ok()){
        ros::spinOnce();
        r.sleep();
    }
    return 0;
}
